web/components/contract-search.tsx

```diff
@@ -103,26 +91,73 @@ export function ContractSearch(props: {
     cardHideOptions,
     highlightOptions,
     headerClassName,
-    useQuerySortLocalStorage,
+    persistPrefix,
     useQuerySortUrlParams,
   } = props
 
-  const [numPages, setNumPages] = useState(1)
-  const [pages, setPages] = useState<Contract[][]>([])
-  const [showTime, setShowTime] = useState<ShowTime | undefined>()
+  const store = safeSessionStorage()
+  const persistAs = (name: string) => {
+    return persistPrefix ? { prefix: persistPrefix, name, store } : undefined
+  }
+
+  const [numPages, setNumPages] = usePersistentState(1, persistAs('numPages'))
```

Comment:
Alternate idea: store persistent state keyed by the url in session storage. This hook would not take any prefix or name params, and simply load and save data based on the url. (Writing from multiple of the same hook at the same url would overwrite the same data.)

If you navigate back to the exact url as before, then it loads the saved data, otherwise, if it loads the same pathname but the query params are different, then it invalidates the cached data. (If it loads a different pathname, the cache is unaffected.)

To use it you'd basically make one object with all the data you want to save for the page. Could make this PR a lot simpler!

The scrolling bit can be independent, and might already work with the top level hook `usePreserveScroll` already discussed.

Reply:
Interesting. I notice a few knobs on this idea:
- It sounds like you are suggesting we manage all the cached state with one hook that has a big blob, instead of calling many hooks with individual values. That alone would clean it up a lot.
- It sounds like you are saying the cache has exactly zero or one entry per page, but it could have more than one if that were useful.
- It sounds like you are saying the cache would invalidate itself if any part of the URL changed, but we could limit it to the path and the query parameters that were actually affecting the relevant state.

I think I agree that this would be simpler and shorter. The main drawback is that we actually would have to use the URL whenever we used this, or else we would have no cache key sitting around. So it might be inappropriate for a lot of the things we currently cache in local storage (whereas as it stands this PR would simplify all those quite nicely.) It's possible, though, that we can naturally make a `useURLState` hook that uses a slightly lower level `usePersistentState` hook and get the best of both worlds.

I'll experiment with this suggestion and these knobs for an hour or two and see what comes out. I like the idea.